# Boucles: Algorithmique de base

Dans le notebook précédent on a vu comment parcourir un intervalle de nombres, et afficher les nombres. On a aussi vu des boucles conditionnelles très simples.

Ici on va voir des cas d'utilisation plus pratiques des boucles inconditionnelles: trouver un élément, compter des éléments, faire une somme ou un produit.

## Trouver un élément dans une liste

Le problème général de trouver un élément dans une liste est très répandu, et sujet à une erreur très classique.

Pour l'instant, comme "liste" on va simplement prendre un intervalle de nombres, qu'on sait maintenant parcourir avec une boucle ```for```.

#### Exemple 6: trouver un multiple de 41
**Question** : Y a-t-il un multiple de 41 dans l'intervalle \[191, 207\]?

Pour répondre à cette question, on va parcourir l'intervalle et chercher un nombre qui répond à cette condition.

La boucle for pour parcourir l'intervalle (on va juste afficher les nombres):

In [6]:
for (int i=191; i<=207; i++){
    System.out.println(i);
}

191
192
193
194
195
196
197
198
199
200
201
202
203
204
205
206
207


Ok, on a maintenant la liste de nombres. Maintenant dans cette liste on va voir s'il y a un multiple de 41. Pour savoir si un nombre est multiple de 41, on utilise l'opérateur modulo: si on divise un nombre par 41 et que le reste est zéro, alors le nombre est multiple de 41. 

Voici le code pour lire un nombre au clavier et tester si il est multiple de 41:

In [2]:
import java.util.Scanner;
Scanner clavier = new Scanner(System.in);

In [3]:
System.out.println("Entrer un nombre:");
int nombre = clavier.nextInt();
if (nombre%41==0){
    System.out.println(nombre +" est multiple de 41");
} else {
    System.out.println(nombre +" n'est pas multiple de 41");
}

Entrer un nombre:
534
534 n'est pas multiple de 41


Pour savoir si un nombre dans la liste au dessus est multiple de 41, on peut tous les tester. Pour cela, on place le test à l'intérieur de la boucle, et on remplace le ```nombre``` dans le test par ```i```, qui prend successivement toutes les valeurs qu'on veut tester:

In [5]:
for (int i=191; i<=207; i++){
    if (i%41==0){
        System.out.println(i +" est multiple de 41");
    } else {
        System.out.println(i +" n'est pas multiple de 41");
    }
}

191 n'est pas multiple de 41
192 n'est pas multiple de 41
193 n'est pas multiple de 41
194 n'est pas multiple de 41
195 n'est pas multiple de 41
196 n'est pas multiple de 41
197 n'est pas multiple de 41
198 n'est pas multiple de 41
199 n'est pas multiple de 41
200 n'est pas multiple de 41
201 n'est pas multiple de 41
202 n'est pas multiple de 41
203 n'est pas multiple de 41
204 n'est pas multiple de 41
205 est multiple de 41
206 n'est pas multiple de 41
207 n'est pas multiple de 41


En lisant ces résultats à l'écran, on peut voir qu'un nombre dans l'intervalle (205) est multiple de 41. Mais dans le contexte d'un programme plus long, on voudrait pouvoir faire d'autres calculs en utilisant ces informations: soit l'information que oui, l'intervalle contient bien un multiple de 41 (information de type booléenne) ou bien l'information que l'élément de la liste qui correspond à la condition est le nombre 205 (information de type numérique).

La première est celle qui a tendance a causer des erreurs.  On définit donc une variable booléenne ```trouvé```, et on voudrait qu'après la boucle ```trouvé``` soit ```true``` si l'intervalle contient un multiple de 41. La première intuition est d'assigner le résultat du "test" dans la boucle à la variable:

In [8]:
boolean trouvé;
for (int i=191; i<=207; i++){
    if (i%41==0){
        trouvé=true;
    } else {
        trouvé=false;
    }
}

if (trouvé){
    System.out.println("L'intervalle contient bien un multiple de 41!");
} else {
    System.out.println("L'intervalle ne contient pas de multiple de 41!");
}

L'intervalle ne contient pas de multiple de 41!


Que se passe-t-il?

En fait si on détaille bien ce qui se passe, pour chaque valeur de i on effectue le test et on assigne un valeur à ```trouvé```. La valeur de ```trouvé``` après la boucle est donc la valeur obtenue lors du *dernier* test (pour ```i = 207```. Et comme 207 n'est pas multiple de 41, la valeur de ```trouvé``` est ```false```.

La solution ici est la suivante:
Il faut initialiser ```trouvé``` à ```false``` (avant la boucle) et seulement modifier ```trouvé``` *si on trouve ce qu'on cherche*, c'est à dire ici, si un des nombres de l'intervalle est multiple de 41.



In [9]:
boolean trouvé=false;
for (int i=191; i<=207; i++){
    if (i%41==0){
        trouvé=true;
    } 
    // on ne met pas de clause else!
}

if (trouvé){
    System.out.println("L'intervalle contient bien un multiple de 41!");
} else {
    System.out.println("L'intervalle ne contient pas de multiple de 41!");
}

L'intervalle contient bien un multiple de 41!


Cette fois, ça fonctionne correctement. Par souci d'efficacité, on peut aussi quitter la boucle une fois que le nombre est trouvé: continuer à parcourir l'intervalle ne changera rien. 

Note: Ceci serait surtout utile si on parcourait une liste très longue; ici on ne verra pas de différence. 

Pour quitter la boucle, on utilise une instruction ```break```:

In [10]:
boolean trouvé=false;
for (int i=191; i<=207; i++){
    if (i%41==0){
        trouvé=true;
        break; // ça y est, on l'a trouvé! on peut quitter la boucle.
    } 
    // on ne met pas de clause else!
}

if (trouvé){
    System.out.println("L'intervalle contient bien un multiple de 41!");
} else {
    System.out.println("L'intervalle ne contient pas de multiple de 41!");
}

L'intervalle contient bien un multiple de 41!


Quelques erreurs classiques:
    * la clause ```else```, comme montré au dessus
    * mettre le ```break``` en dehors du ```if```: on quitte alors la boucle dès le premier passage dans la boucle.

#### Exercice 6: Déterminer si un nombre est premier

Comment détermine-t-on un si un nombre *n* est premier? On cherche dans l'intervalle \[2, *n*/2 \] un diviseur de *n*. Si on en trouve un, alors on sait que *n* n'est pas premier. Si on en trouve un, alors on sait que *n* est premier. 

In [ ]:
Scanner clavier = new Scanner(System.in);
int nombre;
do {
    System.out.println("Entrer un entier supérieur à 2 pour déterminer s'il est premier:");
    nombre = clavier.nextInt();
} while (...); // pour s'assurer que le nombre est supérieur à 2

boolean trouvéDiviseur = ...;
for ( ... ){ // boucle pour chercher des diviseurs dans l'intervalle [2, nombre/2]
    
}

if (...){
    System.out.println(nombre + " est premier");
} else {
    System.out.println(nombre + " n'est pas premier");
}

#### Exemple 7 : trouver un multiple de 41 (suite)

Dans l'exemple précédent on cherchait un multiple de 41 dans un certain intervalle, et l'information qui nous intéressait était: y a-t-il un tel nombre dans l'intervalle? À présent on voudrait non seulement savoir s'il existe, mais s'il existe on veut récupérer sa valeur. 

Comme pour l'information "trouvé", on va créer une variable pour la recevoir (cette fois un entier), et lui donner une valeur dans la condition ```if``` qui indique qu'une valeur appropriée a été trouvée. Ici on nomme cette variable ```multiple```. La seule difficulté est de donner une valeur initiale à ```multiple```: si on ne trouve pas de multiple, alors la variable ne veut rien dire. Ici on l'initialise à 0. Comme on garde aussi la variable ```trouvé```, on n'aura pas de difficulté à interpréter la variable ```multiple```.

In [12]:
boolean trouvé=false;
int multiple =0;
for (int i=191; i<=207; i++){
    if (i%41==0){
        trouvé = true;
        multiple = i;
        break; // ça y est, on l'a trouvé! on peut quitter la boucle.
    } 
    // on ne met pas de clause else!
}

if (trouvé){
    System.out.println("L'intervalle contient bien un multiple de 41: "+ multiple);
} else {
    System.out.println("L'intervalle ne contient pas de multiple de 41!");
}

L'intervalle contient bien un multiple de 41: 205


Parfois, pour simplifier le code (ou dans le contexte d'une fonction, qu'on verra plus tard), on utilise une seule variable pour obtenir la valeur de l'élément cherché, et on utilise une valeur particulière pour le cas où on ne trouve pas d'élément:

In [13]:
int multiple =0;
for (int i=191; i<=207; i++){
    if (i%41==0){
        multiple = i;
        break; // ça y est, on l'a trouvé! on peut quitter la boucle.
    } 
    // on ne met pas de clause else!
}

if (multiple!=0){
    System.out.println("L'intervalle contient bien un multiple de 41: "+ multiple);
} else {
    System.out.println("L'intervalle ne contient pas de multiple de 41!");
}

L'intervalle contient bien un multiple de 41: 205


Ici on initialise ```multiple``` à zéro, et si cette valeur n'a pas changé on suppose que l'élément cherché n'a pas été trouvé. Mais il faut faire attention: 0 est en fait multiple de 41, et si on cherchait des multiples de 41 dans l'intervalle \[-10, 10\], la valeur 0 serait la valeur du multiple trouvé, et non pas une indication qu'on l'a trouvé. Il faudrait alors utiliser une autre valeur initiale, ou simplement utiliser une deuxième variable comme au-dessus.

#### Exercice 7: trouver la racine entière d'un nombre

On veut trouver la racine carrée d'un nombre *n*, si elle est entière. Pour ceci, on va simplement parcourir l'intervalle \[0, n\] et tester chaque nombre pour voir si il donne *n* quand on l'élève au carré. Si et quand on trouve la racine, il faut garder cette valeur pour l'afficher à la fin, comme dans l'exemple 7 quand on trouve le multiple de 41. Si on ne trouve pas de racine entière, on affichera un message indiquant cela. Utiliser le squelette de code suivant comme point de départ:

In [ ]:
Scanner clavier = new Scanner(System.in);
int nombre;
do {
    System.out.println("Entrer un entier positif pour obtenir sa racine carrée:");
    nombre = clavier.nextInt();
} while (...); // pour s'assurer que le nombre est positif

int racine = ...;
for ( ... ){ // boucle pour chercher la racine dans l'intervalle [0, nombre]
    
}

if (racine == ...){
    System.out.println("La racine du nombre n'est pas entière");
} else {
    System.out.println("La racine du nombre est "+ racine);
}

## Compter des éléments

Dans les exemples précédents le but était de trouver un élément dans une liste (un intervalle de nombres, mais ça se généralise à un ensemble quelconque qu'on puisse énumérer). Une extension immédiate est de *compter* le nombre d'élements correspondant à une specification.

Par exemple, au lieu de se demander *si* un intervalle contient un ou plusieurs multiples de 41, on peut se demander *combien* de multiples de 41 l'intervalle contient. 

Comme dans le cas précédent on va introduire une variable pour enregistrer l'information; cette fois le compte sera stocké dans un entier. On appelle ce type de variable un *accumulateur*. Ensuite on parcourt l'intervalle, et le mécanisme pour effectivement compter les éléments sera d'ajouter 1 à la variable à chaque fois qu'on rencontre un élément: c'est comme une façon de compter sur ses doigts ou de faire des traits sur un papier. Enfin, il est clair qu'après avoir trouvé un élément on ne peut pas quitter la boucle.

#### Exemple 8: Compter les multiples de 41 
Comptons les multiples de 41 dans l'intervalle \[128, 305\] (dans l'intervalle précédent, il y en avait un seul):

In [16]:
int compte =0;
for (int i=128; i<=305; i++){
    if (i%41==0){
        compte = compte+1;
    } 
    // on ne met pas de clause else!
}

System.out.println("L'intervalle contient " + compte + " multiples de 41!");


L'intervalle contient 4 multiples de 41!


#### Exercice 8: les multiples de 3 qui se terminent par 1

On considère tous les multiples de 3 inférieurs à 10000. Combien d'entre eux se terminent par le chiffre 1?
Pour le savoir, énumérer les multiples de 3 à l'aide d'une boucle (commencer à zéro et aller de 3 en 3) et compter ceux qui se terminent par le chiffre 1. Pour connaitre le chiffre des unités, utiliser l'opération modulo. Exemple: 438 % 10 = 8.

#### Exercice 9: les multiples de 3 qui se terminent par 1 (bis)
On considère toujours les multiples de 3 inférieurs à 10000. On voudrait savoir quelle *proportion* de ces nombres se termine par 1. Il faut maintenant compter les multiples de 3 et, à part, ceux qui se terminent par 1.

## Faire une somme

Faire la somme de quelques nombres est facile: on écrit ```a + b + c + d``` par exemple. Mais comment faire quand il y a beaucoup de nombres à additionner? Comment, par exemple, faire la somme des nombres de 1 à 100? ou la somme des multiples de 7 (inférieurs à 100)? ou la somme des carrés parfaits (inférieurs à 100)?

La technique est très similaire à celle pour compter des éléments: on les énumère, on définit une variable pour garder la somme, et chaque fois qu'on rencontre un élément, au lieu d'ajouter un au compte, on ajoute la valeur de l'élément à la somme.

#### Exemple 9: La somme des entiers jusqu'à 100:

In [17]:
int somme =0;
for (int i=1; i<=100; i++){
    somme = somme + i;
}
System.out.println("Somme="+somme);

Somme=5050


Ici on a additionné tous les nombres de l'intervalle. Si on additionne seulement des nombres particuliers de cet intervalle (par exemple les multiples de 3 se terminant par le chiffre 1), on peut rajouter une condition pour identifier les éléments à additionner, comme on faisait pour les compter.

#### Exemple 10: Somme des multiples de 3 inférieurs à 100 se terminant par le chiffre 1

In [19]:
int somme3 = 0;
for (int i=0;i<100;i=i+3){
    if(i%10==1){
        somme3 = somme3 + i;
    }
}
System.out.println("Somme des multiples de 3 finissant par 1 = " + somme3);

Somme des multiples de 3 finissant par 1 = 153


#### Exercice 10: Nombres parfaits
Un nombre est dit *parfait* s'il est égal à la somme de ses diviseurs propres (propres = le nombre lui-même exclu). 
Par exemple, 6 est parfait: ses diviseurs propres sont 1,2, et 3, et 1+2+3=6.
Remarquer qu'on inclut 1 comme diviseur.

Comment déterminer si un nombre est parfait? On fait la somme de ses diviseurs et on compare avec le nombre lui-même.
Pour cet exercice, on pourra adapter le code de l'exercice 6: au lieu de chercher un diviseur de *n*, cette fois on veut les lister tous et en faire la somme.

Attention: pour un nombre parfait, on inclut 1 dans la somme, il faut donc commencer la boucle à 1 et non pas à 2.

## Faire un produit

Maintenant qu'on sait faire une somme, est-ce qu'on peut utiliser une technique similaire pour faire un produit? Comme pour une somme, il est facile de faire le produit d'un petit nombre de facteurs: on écrit ```a * b * c * d```.

La réponse est oui: on peut utiliser la même technique. La seule différence est qu'il faut initialiser l'accumulateur à 1 au lieu de l'initialiser à zéro. 

#### Exemple 11: Factorielle
La factorielle d'un nombre *n* est le produit de tous les nombres jusqu'à *n*: *n*! = 1 * 2 * 3 ... * n.

In [20]:
Scanner clavier = new Scanner(System.in);
System.out.println("Entrer un nombre:");
int n = clavier.nextInt();
int produit = 1;
for (int i = 1; i<=n;i++){ 
    produit = produit * i;
}

System.out.println(n + "! = "+produit);


Entrer un nombre:
7
7! = 5040


On vérifie:

In [21]:
1*2*3*4*5*6*7

5040

#### Exercice 11: Produit des diviseurs
Sur le modèle de l'exercice 10, écrire du code pour calculer le *produit*  des diviseurs propres d'un nombre (lu au clavier). 


Note: ceci ne donne pas, en général, la décomposition du nombre en facteurs premiers.
